# Introduction

## Snowden

Edward Joseph Snowden is an American computer professional, a former CIA employee and a former contractor for the Unites States government who copied and leaked classified information from the NSA disclosing numerous global surveillance programs.

Ref: https://en.wikipedia.org/wiki/Edward_Snowden

# Why did I pick Snowden?

Snowden was a topic that was discussed about a lot at one point of time. Was it good or bad what Snowden did? Why is there a movie on him? I heard a lot of controversies about these topics. I too had a mixed reviews.
* I wanted to do a sentiment analysis on this topic to get a general idea on how people think about this.
* Is it positive or negetive?
* A managerial insight could be whether this deed should be taken seriously by the govenment or not? Or is it right or wrong for the public to know this information.

# How did I perform the sentiment analysis?

I am very very new to Python, so to start this I had to read a lot about how to code at first and then about what and how to do the sentiment analysis.
The best part about this Assignment is that many have already done sentiment analysis before which helped me a lot to acquire the data i desired. 
Out of the many links that I read, I liked the below links
* http://adilmoujahid.com/posts/2014/07/twitter-analytics/
* https://gist.github.com/yanofsky/5436496
* https://dev.to/rodolfoferro/sentiment-analysis-on-trumpss-tweets-using-python-

But for the extraction of tweets I followed the first link http://adilmoujahid.com/posts/2014/07/twitter-analytics/

#### Steps I took to acquire the data
First: I signed up with twitter and created an application.
Second: I pasted the entire code from that link in Jupyter notebook (which included the secrect keys and access tokens) to see if it works, but it resulted in many errors.
Second: I installed Python in my windows 10 system and created a python file with the code and ran it in my command line, but this resulted in the same errors.
Third: I analysed the code line by line to figure out mistakes and started correcting it by trial and error method. Alas! it started working.
Fourth: I then entered the tag "Snowden" in the code and ran the python file while i piped the output of the file into a text file on command line.
I let this run for a few hours and it resulted in a text file with JSON output tweets (with metadata) of size 2.06 GB.
Fifth: Since the size was huge to upload to jupyter notebook, i re-ran the python file and this time i stopped running it (Ctrl+C) about 15 minutes and this resulted in a text file of close to 500 tweets.

#### Positive, Negetive & Stopwords
Althought I have given GRE and learned words from Barons and Princetons list, I could not come up with positive, negetive and stopwords by myself. Hence, I googled this information and I found a list of these words that people have already used for their sentiment analysis. 
I simply pasted those words into a 3 different text files (positive, negetive and stopwords).
##### Finally I was ready with my data that I needed to complete this sentiment analysis

#### Below is the code & steps I took to check whether the few tweets are inclining towards positive or negetive state
As I am new to Python, I had to google codes based on my logic and I also had a lot of help from my friends. In doing so I learned a good deal about python.


In [46]:
# print all the outputs in a cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [47]:
t = []
b = open('twitter_data_snowden1.txt', 'r')
for c in b:
    c = c.strip()
    c = c.lower()
    c = c.replace('!','').replace(',','').replace('.','')
    for w in c.split():
        t.append(w)
b.close()


First: Wanted to clean the data and remove some characters and transform it to a list format. This would help in the comparision with the positive, negetive and stopwords.
To confirm if the text file got redered properly, I checked the lengeth of the list below.

In [58]:
len(t)

151587

In [49]:
neg = {}
b = open('neg.txt', 'r')
for c in b:
    c = c.strip()
    neg[c] = 0
b.close()

In [50]:
pos = {}
b = open('pos.txt', 'r')
for c in b:
    c = c.strip()
    pos[c] = 0
b.close()

In [51]:
stop = {}
b = open('stop.txt', 'r')
for c in b:
    c = c.strip()
    stop[c] = 0
b.close()

Second: A precautionary messure to remove the extra spaces in the positive, negetive and stopwords text file while converting them to lists.

In [62]:
p = 0
n = 0
s = 0
plist = []
nlist = []
slist = []
for c in t:
    if c in stop:
        slist.append(c)
        s = s+1
        continue
    if c in pos:
        plist.append(c)
        p = p + 1
    elif c in neg:
        nlist.append(c)
        n = n + 1
        
# t - list with tweets
# stop - list with stopwords
# pos - list with positive words
# neg - list with negetive words

In [70]:
import pandas as pd
poslist = pd.DataFrame(plist)
poslist.to_csv('poslist.csv',index=False,header=False)

In [71]:
import pandas as pd
neglist = pd.DataFrame(nlist)
neglist.to_csv('neglist.csv',index=False,header=False)

In [72]:
import pandas as pd
stoplist = pd.DataFrame(slist)
stoplist.to_csv('stoplist.csv',index=False,header=False)

Third: I used a for-loop to compare the positive, negetive and stopword list with the list of tweets on #Snowden.

Fourth: I created and included additional lists like slist, plist and nlist to store the words that match according to the forloop. I then exported them to CSV to view them later.

Fifth: Below is the code lines where I have found the number of positive, negetive words used and their ratio.

# Observation

In [67]:
p 
n 
s


2927

2498

38611

In [64]:
ratio = p/n

ratio

1.1717373899119294

From the above code we observe that 
* Out of 151587, 2927 words are positive, 2498 words are negetive and 38611 words are stop words.
* The ratio of positive to negetive words is 1.172 which is positive. 

# Interpretation & Conclusion
From the positive ratio found, we can see that most of the words belong to the positive category. This would say that the overall sentiment is positive.

# Managerial Insight I could offer
Despite the ratio being positive there is a lot of tweets out there of the negetive kind. And looking at the overall analysis being positive, posts should be put out supporting Snowden. We could instigate the government to bring down the severity of the action that would be taken against Snowden so that he could go home!.

The above is the managerial insight I could offer only based on the ratio. But the main question here is, is this ratio or is the analysis enough to come up with a managerial insight like this?

# Roadmap for future
* To analyse all the tweets (2.06 GB) and not just part of it.
* To compare sentences instead of just words as negetive/positive because before a negetive word there could be a "not" indicating that its not a negetive statement.
* To be very efficient and provide an accurate Managerial Insight on this topic